google/cloud/storage/internal/grpc_split_write_object_data.cc

```diff
@@ -54,26 +41,35 @@ std::string SplitObjectWriteData<std::string>::Next() {
   return result;
 }
 
-SplitObjectWriteData<absl::Cord>::SplitObjectWriteData(absl::string_view buffer)
-    : cord_(buffer) {}
-
-SplitObjectWriteData<absl::Cord>::SplitObjectWriteData(
-    google::cloud::storage::internal::ConstBufferSequence const& buffers) {
-  for (auto const& b : buffers) {
-    cord_.Append(absl::string_view{b.data(), b.size()});
-  }
-}
-
-bool SplitObjectWriteData<absl::Cord>::Done() const {
-  return offset_ >= cord_.size();
-}
-
+template <>
 absl::Cord SplitObjectWriteData<absl::Cord>::Next() {
   auto constexpr kMax = static_cast<std::size_t>(
       google::storage::v2::ServiceConstants::MAX_WRITE_CHUNK_BYTES);
-  auto const n = (std::min)(kMax, cord_.size() - offset_);
-  auto result = cord_.Subcord(offset_, n);
-  offset_ += n;
+  absl::Cord result;
+  auto start = offset_;
+  for (auto const& b : buffers_) {
+    if (start >= b.size()) {
+      start -= b.size();
+      continue;
+    }
+    auto n = (std::min)(kMax - result.size(), b.size() - start);
+    if (n != 0) {
+      // We need a container which guarantees the pointer is stable under move
+      // construction. `std::vector<>` does not provide that guarantee. Use
+      // `std::unique_ptr<char[]>`. Do not use `std::make_unique<char[]> as that
+      // will zero-initialize the array, and we are just going to overwrite the
+      // bytes.
+      auto buffer = std::unique_ptr<char[]>(new char[n]);
+      std::copy_n(b.begin() + start, n, buffer.get());
+      // Get `buffer.get()` before we std::move-from.
+      auto contents = absl::string_view{buffer.get(), n};
+      result.Append(absl::MakeCordFromExternal(
+          contents, [b = std::move(buffer)]() mutable {}));
+    }
+    offset_ += n;
+    start = 0;
+    if (result.size() >= kMax) return result;
```

Comment:
Nit: I think I would make both occurrences of this say `break` instead of `return`.

Reply:
Ack.  I can clean this up tomorrow.